src/algorithms/l0_smooth.jl

```diff
@@ -0,0 +1,183 @@
+
+# Examples
+
+You can use the default arguments for `L0Smooth`, and then use `smooth` to apply 
+the `AbstractImageSmoothAlgorithm`.
+
+```julia
+using TestImages
+using ImageSmooth
+
+img = testimage("cameraman")
+
+fₛ = L0Smooth() # using default arguements
+imgₛ = smooth(img, fₛ)
+```
+
+Manually setting the arguements is also available:
+
+```julia
+fₛ = L0Smooth(λ=0.0015, κ=1.05, βmax=2e5) # manually set the arguments
+imgₛ = smooth(img, fₛ)
+```
+
+See also [`smooth!`](@ref) for in-place operation.
+
+# References
+
+[1] Xu, L., Lu, C., Xu, Y., & Jia, J. (2011, December). Image smoothing via L 0 gradient minimization. In Proceedings of the 2011 SIGGRAPH Asia conference (pp. 1-12). [DOI:10.1145/2024156.2024208](https://doi.org/10.1145/2024156.2024208)
+
+"""
+struct L0Smooth <: AbstractImageSmoothAlgorithm
+    """smoothing weight"""
+    λ::Float64
+    """iteration rate"""
+    κ::Float64
+    """the upper bound of automatically adapting parameter β which controls the iteraiton times"""
+    βmax::Float64
+
+    function L0Smooth(λ, κ, βmax)
+        λ > zero(λ) || throw(ArgumentError("smoothing weight λ should be positive"))
+        κ > one(κ) || throw(ArgumentError("iteration rate κ should be > 1.0"))
+        βmax > 1e4 || throw(ArgumentError("the iteration upper bound βmax should be > 1e4"))
+        new(λ, κ, βmax)
+    end
+end
+
+L0Smooth(; λ::Float64=2e-2, κ::Float64=2.0, βmax::Float64=1e5) = L0Smooth(λ, κ, βmax)
+
+
+function (f::L0Smooth)(out::AbstractArray{<: Number},
+                       img::AbstractArray{<: Number})
+    𝑆 = float64.(img)
+    𝜆 = f.λ # smoothing weight
+    𝜅 = f.κ # iteration rate
+    𝛽max = f.βmax # upper bound of 𝛽
+    𝛽 = 2 * 𝜆 # define 𝛽₀
+    ∂₁ = [1 -1]
+    ∂₂ = [1, -1]
+    D, N, M = size(𝑆)
+    sizeI2D = (N, M)
+    sizeI2D_t = (M, N)
+    # Return the frequency-space representation of `ℱ∂₁` and `ℱ∂₂`.
+    ℱ∂₁ = freqkernel(centered(∂₁), sizeI2D)
+    ℱ∂₂ = transpose(freqkernel(centered(transpose(∂₂)), sizeI2D_t))
+    ℱ𝐼 = fft(𝑆, (2, 3))
+    Denormin = @. abs(ℱ∂₁)^2 + abs(ℱ∂₂)^2
+    Denormin = reshape(Denormin, 1, size(Denormin)...)
+    if D > 1
+        Denormin = repeat(Denormin, inner=(1, 1, 1), outer=(D, 1, 1))
+    end
+
+    𝛥₁𝑆 = similar(𝑆)
+    𝛥₂𝑆 = similar(𝑆)
+    𝛻₁ℎ = similar(𝑆)
+    𝛻₂𝑣 = similar(𝑆)
+
+    Normin = similar(ℱ𝐼) 
+    t¹ = trues(1, N, M)
+    t³ = trues(D, N, M)
+    ℱ𝑆 = similar(ℱ𝐼)
+
+    while 𝛽 < 𝛽max
+        # Computing (ℎ, 𝑣) via solving equation (9) in [1]
+        # We get the solution (12) in [1] through following process
+        # Use (𝛥₁𝑆, 𝛥₂𝑆) to demonstrate (ℎ, 𝑣) for convenience
+        forwarddiff!(𝛥₁𝑆, 𝑆, dims = 3)
+        forwarddiff!(𝛥₂𝑆, 𝑆, dims = 2)
+
+        # For each pixel 𝑝 in 𝑆
+        # (ℎₚ, 𝑣ₚ) = (0, 0), while (𝛥₁𝑆ₚ^2 + 𝛥₂𝑆ₚ^2) < λ / 𝛽
+        # (ℎₚ, 𝑣ₚ) = (𝛥₁𝑆ₚ, 𝛥₂𝑆ₚ), otherwise
+        if D > 1
+            t¹ .= sum((𝛥₁𝑆.^2 .+ 𝛥₂𝑆.^2), dims=1) .< 𝜆 / 𝛽
+            t³ .= repeat(t¹, inner=(1, 1, 1), outer=(D, 1, 1))

```

Comment:
These two lines also introduce unnecessary memory allocations:

1. `(𝛥₁𝑆.^2 .+ 𝛥₂𝑆.^2)` creates an intermediate array, and then pass into `sum`
2. `repeat` creates a new array

For 1 there are two options:

- preallocate a buffer to store the result of `sum` and reuse it in the while loop
- use another norm definition: `sum(abs2, 𝛥₁𝑆) .+ sum(abs2, 𝛥₂𝑆) .< 𝜆 / 𝛽`

For 2, the only way is to pre-allocate a buffer.

Reply:
I'll take a try in the next PR.